mm2src/lp_ordermatch/simple_market_maker.rs

```diff
@@ -0,0 +1,530 @@
+    InternalError(String),
+}
+
+#[derive(Debug, Deserialize, Display, Serialize, SerializeErrorType)]
+#[serde(tag = "error_type", content = "error_data")]
+pub enum StartSimpleMakerBotError {
+    #[display(fmt = "The bot is already started")]
+    AlreadyStarted,
+    #[display(fmt = "Invalid bot configuration")]
+    InvalidBotConfiguration,
+    #[display(fmt = "Transport error: {}", _0)]
+    Transport(String),
+    #[display(fmt = "Internal error: {}", _0)]
+    InternalError(String),
+}
+
+#[derive(Debug)]
+pub enum PriceServiceRequestError {
+    HttpProcessError(String),
+}
+
+impl From<std::string::String> for PriceServiceRequestError {
+    fn from(error: String) -> Self { PriceServiceRequestError::HttpProcessError(error) }
+}
+
+impl From<std::str::Utf8Error> for PriceServiceRequestError {
+    fn from(error: Utf8Error) -> Self { PriceServiceRequestError::HttpProcessError(error.to_string()) }
+}
+
+impl HttpStatusCode for StartSimpleMakerBotError {
+    fn status_code(&self) -> StatusCode {
+        match self {
+            StartSimpleMakerBotError::AlreadyStarted | StartSimpleMakerBotError::InvalidBotConfiguration => {
+                StatusCode::BAD_REQUEST
+            },
+            StartSimpleMakerBotError::Transport(_) | StartSimpleMakerBotError::InternalError(_) => {
+                StatusCode::INTERNAL_SERVER_ERROR
+            },
+        }
+    }
+}
+
+impl HttpStatusCode for StopSimpleMakerBotError {
+    fn status_code(&self) -> StatusCode {
+        match self {
+            // maybe bad request is not adapted for the first errors.
+            StopSimpleMakerBotError::AlreadyStopped | StopSimpleMakerBotError::AlreadyStopping => {
+                StatusCode::BAD_REQUEST
+            },
+            StopSimpleMakerBotError::Transport(_) | StopSimpleMakerBotError::InternalError(_) => {
+                StatusCode::INTERNAL_SERVER_ERROR
+            },
+        }
+    }
+}
+
+struct TradingPair {
+    base: String,
+    rel: String,
+}
+
+impl TradingPair {
+    pub fn new(base: String, rel: String) -> TradingPair { TradingPair { base, rel } }
+
+    pub fn as_combination(&self) -> String { self.base.clone() + "/" + self.rel.clone().as_str() }
+}
+
+pub async fn tear_down_bot(ctx: MmArc) {
+    let simple_market_maker_bot_ctx = TradingBotContext::from_ctx(&ctx).unwrap();
+    {
+        let mut trading_bot_cfg = simple_market_maker_bot_ctx.trading_bot_cfg.lock().await;
+        // todo: check if clear is adapted, if i understand its keep the memory allocated for later usage.
+        trading_bot_cfg.clear();
+    }
+    // todo: cancel all pending orders
+}
+
+// This function check if coin is enabled and if check_balance is set to true and the balance is non-zero it's returned for later usage
+// e.g i want to use 50% of my KMD balance instead of max, will be used later.
+async fn coin_find_and_checks(
+    ticker: String,
+    key_trade_pair: String,
+    check_balance: bool,
```

Comment:
I've tried this approach but the `coin.get_non_zero_balance().await?` is quite challenging because we have to return a `MmError<BalanceError>` and then I need to convert it to `MmError<OrderProcessingError>`

And we cannot define `from` traits for a type defined in another crate - any tips?

Reply:
`get_non_zero_balance` should return `GetNonZeroBalanceError` as outlined in the review below. This way you won't have to define `From` for the type from an external crate.